exercises/concept/take-a-number-deluxe/.meta/exemplar.ex

```diff
@@ -0,0 +1,82 @@
+defmodule TakeANumberDeluxe do
+  use GenServer
+
+  # Client API
+
+  def start_link(init_arg) do
+    GenServer.start_link(__MODULE__, init_arg)
+  end
+
+  def report_state(machine) do
+    GenServer.call(machine, :report_state)
+  end
+
+  def queue_new_number(machine) do
+    GenServer.call(machine, :queue_new_number)
+  end
+
+  def serve_next_queued_number(machine, priority_number \\ nil) do
+    GenServer.call(machine, {:serve_next_queued_number, priority_number})
+  end
+
+  def reset_state(machine) do
+    GenServer.cast(machine, :reset_state)
+  end
+
+  # Server callbacks
+
+  @impl true
+  def init(init_arg) do
+    min_number = Keyword.get(init_arg, :min_number)
+    max_number = Keyword.get(init_arg, :max_number)
+
+    case TakeANumberDeluxe.State.new(min_number, max_number) do
+      {:ok, state} -> {:ok, state}
+      {:error, error} -> {:stop, error}
+    end
+  end
+
+  @impl true
+  def handle_call(:report_state, _from, state) do
+    {:reply, state, state}
+  end
+
+  @impl true
+  def handle_call(:queue_new_number, _from, state) do
+    case TakeANumberDeluxe.State.queue_new_number(state) do
+      {:ok, new_number, new_state} ->
+        {:reply, {:ok, new_number}, new_state}
+
+      {:error, error} ->
+        {:reply, {:error, error}, state}
+    end
+  end
+
+  @impl true
+  def handle_call({:serve_next_queued_number, priority_number}, _from, state) do
+    case TakeANumberDeluxe.State.serve_next_queued_number(state, priority_number) do
+      {:ok, next_number, new_state} ->
+        {:reply, {:ok, next_number}, new_state}
+
+      {:error, error} ->
+        {:reply, {:error, error}, state}
+    end
+  end
+
+  @impl true
+  def handle_cast(:reset_state, state) do
+    {:ok, state} = TakeANumberDeluxe.State.new(state.min_number, state.max_number)
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info({:take_a_number, sender_pid}, state) do
+    send(sender_pid, {:error, :basic_model_message_received_by_deluxe_model_server})
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info(_, state) do
+    {:noreply, state}
+  end
```

Comment:
Running `use GenServer` already creates a catch-all `handle_info`, so I had to get creative to push students to implement it themselves and actually be able to assert they did. At first I wanted just a normal catch-all that calls `IO.inspect` but I couldn't figure out how to assert on the IO. The normal capture IO in ExUnit doesn't work because the IO happens in a different process.

So I came up with this, and I'm going to justify it with something like: "when testing the initial prototype of the new machine, somebody did something very weird in the client code and ended up sending the old messages meant for the basic Take-A-Number machine to the new GenServer based one instead. They were really annoyed that they didn't receive any responses or any errors, and asked you to add this response for them so they realize what's going on the next time". A bit far-fetched but I was out of ideas.

Reply:
If you want another idea for justifying `handle_info`, elixir in Action uses the example of sending messages at regular intervals
```elixir
def init(_) do
  :timer.send_interval(5000, :cleanup)
  ...
end
```
You could say that the number queue should be reset once a day.

Reply:
Another way to push students to implement the `handle_info` callback is to use GenServer's own [timeout][genserver-timeout]. You can say in history that for "economy" reasons the machine should be turned off after a period of inactivity.

Example:

```elixir
@timeout 1500

def init(_), do: {:ok, ..., @timeout}

def handle_call(..., ..., ...), do: {:reply, ..., ..., @timeout}
def handle_cast(..., ...), do: {:noreply, ..., @timeout}
def handle_info(:timeout, ...), do: exit(:normal) # energy_is_expensive
```

That way, if the student doesn't implement this callback, the process will keep running and the tests should fail, since the _match all_ `handle_info` just calls a `:logger.error`

```elixir
Process.sleep(1000)
assert Process.alive?(pid)
Process.sleep(1000)
refute Process.alive?(pid)
```

_This behavior should be easy to test, but these timers in the tests scares me_

[genserver-timeout]: https://hexdocs.pm/elixir/GenServer.html#module-timeouts

Reply:
@Tyyagoo That sounds brilliant! I totally forgot about the "timeouts" feature. I will try to incorporate your idea.